services/src/main/java/org/keycloak/protocol/saml/SamlService.java

```diff
@@ -732,14 +888,168 @@ public AuthenticationSessionModel getOrCreateLoginSessionForIdpInitiatedSso(Keyc
+
+        logger.debug("Received artifactResolve message for artifact " + artifactResolveMessage.getArtifact() + "\n" +
+                "Message: \n" + DocumentUtil.getDocumentAsString(samlDoc.getSamlDocument()));
+
+        String issuer = artifactResolveMessage.getIssuer().getValue();
+        ClientModel client = realm.getClientByClientId(issuer);
+
+        if (client == null) {
+            throw new ProcessingException(Errors.CLIENT_NOT_FOUND);
+        }
+        if (!client.isEnabled()) {
+            throw new ProcessingException(Errors.CLIENT_DISABLED);
+        }
+        if (client.isBearerOnly()) {
+            throw new ProcessingException(Errors.NOT_ALLOWED);
+        }
+        if (!client.isStandardFlowEnabled()) {
+            throw new ProcessingException(Errors.NOT_ALLOWED);
+        }
+        try {
+            SamlProtocolUtils.verifyDocumentSignature(client, samlDoc.getSamlDocument());
+        } catch (VerificationException e) {
+            SamlService.logger.error("request validation failed", e);
+            throw new ProcessingException(Errors.INVALID_SIGNATURE);
+        }
+
+        String artifactResponse = getArtifact(client, artifactResolveMessage.getArtifact());
+        Document artifactResponseDocument = DocumentUtil.getDocument(artifactResponse);
+
+        Soap.SoapMessageBuilder messageBuilder = Soap.createMessage();
+        messageBuilder.addToBody(artifactResponseDocument);
+
+        logger.debug("Sending artifactResponse message for artifact " + artifactResolveMessage.getArtifact() + "\n" +
+                "Message: \n" + artifactResponse);
+
+        return messageBuilder.build();
+    }
+
+    /**
+     * Retrieves an artifact from a client's cache. The method does this by going through the UserSessions attached to the
+     * client.
+     * @param client the clients session for which to query the artifact
+     * @param artifact an artifact value
+     * @return the document containing the corresponding artifact response, which was saved to the session
+     * @throws ParsingException
+     * @throws ConfigurationException
+     * @throws ProcessingException
+     */
+    private String getArtifact(ClientModel client, String artifact) throws ParsingException, ConfigurationException, ProcessingException {
+        List<UserSessionModel> userSessions = session.sessions().getUserSessions(realm, client);
+        for (UserSessionModel userSession: userSessions) {
+            AuthenticatedClientSessionModel clientSession = userSession.getAuthenticatedClientSessionByClient(client.getId());
+            if (clientSession != null) {
+                String response = clientSession.getNote(artifact);
+                if (response != null && ! response.isEmpty()) {
+                    clientSession.removeNote(artifact);
+                    return response;
+                }
+            }
+        }
+        //check if it is the logout message
```

Comment:
The change of the deferred model of user session removal is unfortunately not justifiable in this case. However there is way to build the SAML art in a way that it contains all the information needed to reconstruct logout response. SAML art has to be prepared so that it is not forgerable. This can be done by actually creating a JWT token (yes, the same from OIDC) upon creating the Artifact response with limited time validity. This would be validated there via `TokenVerifier` and based on its values, the LogoutResponse would be sent (or an error in case it would be invalid).

If you prefer, this part can be written by myself, feel free to let me know.